src/ucsi/ppm/state_machine.rs

```diff
@@ -0,0 +1,162 @@
+pub enum State {
+    /// Idle(notification enabled)
+    Idle(bool),
+    /// Busy(notification enabled)
+    Busy(bool),
+    /// Processing Command,
+    ProcessingCommand,
+    /// Waiting for command complete ack
+    WaitForCommandCompleteAck,
+    /// Waiting for async event ack
+    WaitForAsyncEventAck,
+}
+
+/// Inputs to the PPM state machine
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+#[cfg_attr(feature = "defmt", derive(defmt::Format))]
+pub enum Input {
+    /// Notification enabled
+    NotificationEnabled,
+    /// PPM reset
+    Reset,
+    /// Current command cancelled
+    Cancel,
+    /// Command completed immediately
+    CommandImmediate,
+    /// Command did not complete immediately
+    CommandAsync,
+    /// Command completed,
+    CommandCompleted,
+    /// Command complete ack received from OPM
+    CommandCompleteAck,
+    /// Async event ack received from OPM
+    AsyncEventAck,
+    /// External busy status changed
+    BusyChanged,
+    /// There's a pending async event
+    PendingAsyncEvent,
+}
+
+/// Outputs from the PPM state machine
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+#[cfg_attr(feature = "defmt", derive(defmt::Format))]
+pub enum Output {
+    /// Notify OPM that command completed
+    OpmNotifyCommandComplete,
+    /// Notify OPM of async event
+    OpmNotifyAsyncEvent,
+    /// Notify OPM of PPM reset
+    OpmNotifyReset,
+    /// Notify OPM that PPM is busy
+    OpmNotifyBusy,
+}
+
+/// Attempted transition that is not allowed by the state machine
+#[derive(Copy, Clone, Debug)]
+#[cfg_attr(feature = "defmt", derive(defmt::Format))]
+pub struct InvalidTransition {
+    /// The current state of the state machine
+    pub state: State,
+    /// The input that was attempted
+    pub input: Input,
+}
+
+#[cfg_attr(doc, aquamarine::aquamarine)]
+/// include_mmd!("docs/ucsi/ppm_state_machine.mmd")
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+#[cfg_attr(feature = "defmt", derive(defmt::Format))]
+pub struct StateMachine {
+    state: State,
+}
+
+impl StateMachine {
+    /// Create a new state machine in the Idle(false) state
+    pub const fn new() -> Self {
+        StateMachine {
+            state: State::Idle(false),
+        }
+    }
+
+    /// Returns the current state
+    pub fn state(&self) -> State {
+        self.state
+    }
+
+    /// Transition the state machine based on the input and return the output to the OPM if any.
+    pub fn consume(&mut self, input: Input) -> Result<Option<Output>, InvalidTransition> {
+        use Input::*;
+        use Output::*;
+        use State::*;
+
+        let (next_state, output) = match (self.state, input) {
+            // Idle(false) transitions
+            (Idle(false), NotificationEnabled) => (Idle(true), None),
+            (Idle(false), BusyChanged) => (Busy(false), None),
+            (Idle(false), CommandImmediate) => (Idle(false), None),
+            (Idle(false), CommandAsync) => (Idle(false), None),
```

Comment:
Any desire to combine branches?
```suggestion
            (Idle(false), CommandImmediate | CommandAsync) => (Idle(false), None),
```

Several opportunities for combining